File: clients/node/src/consumer.ts
import { ConsumerOptions, BatchOptions } from './options'
import { Message, buildMessage } from './message'
import { sleep } from './internal/sleep'

export type MessageHandler = (msg: Message) => Promise<void>
export type BatchHandler = (messages: Message[]) => Promise<void>

const BACKOFF_START_MS = 500
const BACKOFF_MAX_MS = 30_000

function nextBackoff(current: number): number {
  return Math.min(current * 2, BACKOFF_MAX_MS)
}

function callbackToPromise<T>(
  fn: (cb: (err: Error | null, res: T) => void) => void,
): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    fn((err, res) => {
      if (err) reject(err)
      else resolve(res)
    })
  })
}

// ProtoTimestamp mirrors the shape @grpc/proto-loader produces for google.protobuf.Timestamp.
interface ProtoTimestamp {
  seconds: string | number
  nanos: number
}

function protoTimestampToDate(ts: ProtoTimestamp | null | undefined): Date {
  if (!ts) return new Date(0)
  const secs = typeof ts.seconds === 'string' ? parseInt(ts.seconds, 10) : ts.seconds
  return new Date(secs * 1000 + Math.floor(ts.nanos / 1_000_000))
}

export interface RawMessage {
  id: string
  topic: string
  payload: Buffer
  metadata: Record<string, string>
  createdAt: ProtoTimestamp | null | undefined
  retryCount: number
}

export interface SubscribeStream {
  on(event: 'data', handler: (msg: RawMessage) => void): this
  on(event: 'error', handler: (err: Error) => void): this
  on(event: 'end', handler: () => void): this
  cancel(): void
}

export interface ConsumerStub {
  subscribe(request: {
    topic: string
    visibilityTimeoutSeconds?: number
    consumerGroup?: string
  }): SubscribeStream

  batchDequeue(
    request: { topic: string; count: number; visibilityTimeoutSeconds?: number; consumerGroup?: string },
    callback: (err: Error | null, response: { messages: RawMessage[] }) => void,
  ): void

  ack(request: { id: string; consumerGroup?: string }, callback: (err: Error | null) => void): void
  nack(request: { id: string; error: string; consumerGroup?: string }, callback: (err: Error | null) => void): void
}

export class Consumer {
  private readonly topic: string
  private readonly concurrency: number
  private readonly visibilityTimeoutSeconds: number | undefined
  private readonly signal: AbortSignal | undefined
  private readonly consumerGroup: string

  constructor(
    private readonly stub: ConsumerStub,
    topic: string,
    options?: ConsumerOptions,
  ) {
    this.topic = topic
    this.concurrency = options?.concurrency ?? 1
    this.visibilityTimeoutSeconds = options?.visibilityTimeoutSeconds
    this.signal = options?.signal
    this.consumerGroup = options?.consumerGroup ?? ''
  }

  async consume(handler: MessageHandler): Promise<void> {
    let backoff = BACKOFF_START_MS

    while (!this.signal?.aborted) {
      const req: { topic: string; visibilityTimeoutSeconds?: number; consumerGroup: string } = {
        topic: this.topic,
        consumerGroup: this.consumerGroup,
      }
      if (this.visibilityTimeoutSeconds !== undefined) {
        req.visibilityTimeoutSeconds = this.visibilityTimeoutSeconds
      }

      let stream: SubscribeStream
      try {
        stream = this.stub.subscribe(req)
      } catch (err) {
        if (this.signal?.aborted) return
        console.error(`queue-ti consumer: subscribe error (retrying in ${backoff}ms):`, err)
        try {
          await sleep(backoff, this.signal)
        } catch {
          return
        }
        backoff = nextBackoff(backoff)
        continue
      }

      const cleanExit = await this.drainStream(stream, handler)

      if (cleanExit) {
        backoff = BACKOFF_START_MS
      }

      if (this.signal?.aborted) return

      if (!cleanExit) {
        try {
          await sleep(backoff, this.signal)
        } catch {
          return
        }
        backoff = nextBackoff(backoff)
      }
    }
  }

  private drainStream(
    stream: SubscribeStream,
    handler: MessageHandler,
  ): Promise<boolean> {
    // Semaphore: track active handler slots via a counter + a queue of resolvers
    // waiting for a free slot.
    let activeCount = 0
    const waitQueue: Array<() => void> = []

    const acquireSlot = (): Promise<void> => {
      if (activeCount < this.concurrency) {
        activeCount++
        return Promise.resolve()
      }
      return new Promise<void>((resolve) => waitQueue.push(resolve))
    }

    const releaseSlot = (): void => {
      const next = waitQueue.shift()
      if (next) {
        next()
      } else {
        activeCount--
      }
    }

    return new Promise<boolean>((resolve) => {
      let settled = false
      // Track all in-flight handler promises so we can await them before resolving.
      const inFlight: Promise<void>[] = []

      const settle = (cleanExit: boolean): void => {
        if (settled) return
        settled = true
        // Wait for all in-flight handlers before resolving so we don't drop messages.
        void Promise.allSettled(inFlight).then(() => resolve(cleanExit))
      }

      if (this.signal) {
        this.signal.addEventListener('abort', () => {
          stream.cancel()
          settle(true)
        }, { once: true })
      }

      stream.on('error', (err) => {
        if (this.signal?.aborted) {
          settle(true)
          return
        }
        console.error('queue-ti consumer: stream error (will reconnect):', err)
        settle(false)
      })

      stream.on('end', () => settle(true))

      stream.on('data', (raw: RawMessage) => {
        const msg = this.rawToMessage(raw)

        const handlerPromise = acquireSlot().then(async () => {
          try {
            await this.dispatch(msg, handler)
          } finally {
            releaseSlot()
          }
        })

        inFlight.push(handlerPromise)
      })
    })
  }

  private async dispatch(msg: Message, handler: MessageHandler): Promise<void> {
    let threw = false
    let thrownReason = 'unknown error'
    try {
      await handler(msg)
    } catch (err) {
      threw = true
      thrownReason = err instanceof Error ? err.message : String(err)
    }

    if (!threw) {
      try {
        await msg.ack()
      } catch (err) {
        if (!this.signal?.aborted) {
          console.error(`queue-ti consumer: ack failed for message ${msg.id}:`, err)
        }
      }
      return
    }

    try {
      await msg.nack(thrownReason)
    } catch (err) {
      if (!this.signal?.aborted) {
        console.error(`queue-ti consumer: nack failed for message ${msg.id}:`, err)
      }
    }
  }

  async consumeBatch(options: BatchOptions, handler: BatchHandler): Promise<void> {
    let backoff = BACKOFF_START_MS

    while (!this.signal?.aborted) {
      const req: { topic: string; count: number; visibilityTimeoutSeconds?: number; consumerGroup: string } = {
        topic: this.topic,
        count: options.batchSize,
        consumerGroup: this.consumerGroup,
      }
      if (options.visibilityTimeoutSeconds !== undefined) {
        req.visibilityTimeoutSeconds = options.visibilityTimeoutSeconds
      }

      let messages: RawMessage[]
      try {
        messages = await this.batchDequeue(req)
      } catch (err) {
        if (this.signal?.aborted) return
        console.error(`queue-ti consumer: batchDequeue error (retrying in ${backoff}ms):`, err)
        try {
          await sleep(backoff, this.signal)
        } catch {
          return
        }
        backoff = nextBackoff(backoff)
        continue
      }

      if (messages.length === 0) {
        try {
          await sleep(backoff, this.signal)
        } catch {
          return
        }
        backoff = nextBackoff(backoff)
        continue
      }

      backoff = BACKOFF_START_MS

      const batch = messages.map((raw) => this.rawToMessage(raw))
      try {
        await handler(batch)
      } catch (err) {
        if (!this.signal?.aborted) {
          console.error('queue-ti consumer: batch handler error:', err)
        }
      }
    }
  }

  private batchDequeue(req: {
    topic: string
    count: number
    visibilityTimeoutSeconds?: number
  }): Promise<RawMessage[]> {
    return callbackToPromise<{ messages: RawMessage[] }>(
      (cb) => this.stub.batchDequeue(req, cb),
    ).then((response) => response.messages ?? [])
  }

  private rawToMessage(raw: RawMessage): Message {
    const ackFn = (id: string): Promise<void> =>
      callbackToPromise<void>((cb) => this.stub.ack({ id, consumerGroup: this.consumerGroup }, cb))
        .catch((err: Error) => { throw new Error(`ack message ${id}: ${err.message}`) })

    const nackFn = (id: string, reason: string): Promise<void> =>
      callbackToPromise<void>((cb) => this.stub.nack({ id, error: reason, consumerGroup: this.consumerGroup }, cb))
        .catch((err: Error) => { throw new Error(`nack message ${id}: ${err.message}`) })

    return buildMessage(
      raw.id,
      raw.topic,
      Buffer.isBuffer(raw.payload) ? raw.payload : Buffer.from(raw.payload),
      raw.metadata ?? {},
      protoTimestampToDate(raw.createdAt),
      raw.retryCount ?? 0,
      ackFn,
      nackFn,
    )
  }
}

File: clients/python/queueti/_consumer.py
from __future__ import annotations

import asyncio
import logging
from collections.abc import AsyncIterator, Awaitable, Callable
from datetime import datetime, timezone

import grpc

from queueti._message import Message
from queueti._options import BatchOptions, ConsumerOptions
from queueti.pb import queue_pb2, queue_pb2_grpc

logger = logging.getLogger(__name__)

MessageHandler = Callable[[Message], Awaitable[None]]
BatchHandler = Callable[[list[Message]], Awaitable[None]]

_BACKOFF_START = 0.5
_BACKOFF_MAX = 30.0


def _next_backoff(current: float) -> float:
    return min(current * 2, _BACKOFF_MAX)


def _proto_ts_to_datetime(ts: object) -> datetime:
    seconds = getattr(ts, "seconds", 0) or 0
    nanos = getattr(ts, "nanos", 0) or 0
    return datetime.fromtimestamp(seconds + nanos / 1e9, tz=timezone.utc)


class AsyncConsumer:
    def __init__(
        self,
        stub: queue_pb2_grpc.QueueServiceStub,
        topic: str,
        options: ConsumerOptions | None = None,
    ) -> None:
        self._stub = stub
        self._topic = topic
        opts = options or ConsumerOptions()
        self._concurrency = opts.concurrency
        self._visibility_timeout = opts.visibility_timeout_seconds
        self._consumer_group: str = opts.consumer_group

    async def consume(self, handler: MessageHandler) -> None:
        """Stream messages from the topic, calling handler for each one.

        Runs until cancelled. Auto-acks on success, auto-nacks on handler exception.
        Reconnects with exponential backoff on stream errors.
        """
        backoff = _BACKOFF_START
        while True:
            req = queue_pb2.SubscribeRequest(topic=self._topic, consumer_group=self._consumer_group)
            if self._visibility_timeout is not None:
                req.visibility_timeout_seconds = self._visibility_timeout

            try:
                stream: AsyncIterator[queue_pb2.SubscribeResponse] = self._stub.Subscribe(req)
                clean_exit = await self._drain_stream(stream, handler)
            except asyncio.CancelledError:
                return
            except grpc.RpcError as exc:
                logger.error("queue-ti consumer: subscribe error (retrying in %.1fs): %s", backoff, exc)
                try:
                    await asyncio.sleep(backoff)
                except asyncio.CancelledError:
                    return
                backoff = _next_backoff(backoff)
                continue

            if clean_exit:
                backoff = _BACKOFF_START
            else:
                try:
                    await asyncio.sleep(backoff)
                except asyncio.CancelledError:
                    return
                backoff = _next_backoff(backoff)

    async def _drain_stream(
        self,
        stream: AsyncIterator[queue_pb2.SubscribeResponse],
        handler: MessageHandler,
    ) -> bool:
        sem = asyncio.Semaphore(self._concurrency)
        tasks: set[asyncio.Task[None]] = set()

        async def dispatch(raw: queue_pb2.SubscribeResponse) -> None:
            msg = self._build_message(raw)
            await self._dispatch(msg, handler)

        def _on_done(t: asyncio.Task[None], _sem: asyncio.Semaphore = sem) -> None:
            _sem.release()
            tasks.discard(t)

        try:
            async for raw in stream:
                await sem.acquire()
                task = asyncio.create_task(dispatch(raw))
                tasks.add(task)
                task.add_done_callback(_on_done)
        except asyncio.CancelledError:
            current = asyncio.current_task()
            if current is not None and current.cancelling() > 0:
                # External task.cancel() — abort in-flight handlers quickly.
                for t in tasks:
                    t.cancel()
            # Let tasks finish (or absorb their cancellations) before re-raising.
            await asyncio.gather(*tasks, return_exceptions=True)
            raise  # propagate so consume() exits the while loop
        except grpc.RpcError as exc:
            logger.error("queue-ti consumer: stream error (will reconnect): %s", exc)
            await asyncio.gather(*tasks, return_exceptions=True)
            return False

        await asyncio.gather(*tasks, return_exceptions=True)
        return True

    async def _dispatch(self, msg: Message, handler: MessageHandler) -> None:
        threw = False
        reason = "unknown error"
        try:
            await handler(msg)
        except Exception as exc:
            threw = True
            reason = str(exc)

        if not threw:
            try:
                await self._stub.Ack(queue_pb2.AckRequest(id=msg.id, consumer_group=self._consumer_group))
            except Exception as exc:
                logger.error("queue-ti consumer: ack failed for message %s: %s", msg.id, exc)
            return

        try:
            await self._stub.Nack(queue_pb2.NackRequest(id=msg.id, error=reason, consumer_group=self._consumer_group))
        except Exception as exc:
            logger.error("queue-ti consumer: nack failed for message %s: %s", msg.id, exc)

    async def consume_batch(self, options: BatchOptions, handler: BatchHandler) -> None:
        """Poll batches from the topic, calling handler for each batch.

        Runs until cancelled. Handler is responsible for acking/nacking messages.
        Backs off when the queue is empty.
        """
        backoff = _BACKOFF_START
        while True:
            group = options.consumer_group or self._consumer_group
            req = queue_pb2.BatchDequeueRequest(topic=self._topic, count=options.batch_size, consumer_group=group)
            if options.visibility_timeout_seconds is not None:
                req.visibility_timeout_seconds = options.visibility_timeout_seconds

            try:
                resp: queue_pb2.BatchDequeueResponse = await self._stub.BatchDequeue(req)
            except asyncio.CancelledError:
                return
            except grpc.RpcError as exc:
                logger.error("queue-ti consumer: batchDequeue error (retrying in %.1fs): %s", backoff, exc)
                try:
                    await asyncio.sleep(backoff)
                except asyncio.CancelledError:
                    return
                backoff = _next_backoff(backoff)
                continue

            if not resp.messages:
                try:
                    await asyncio.sleep(backoff)
                except asyncio.CancelledError:
                    return
                backoff = _next_backoff(backoff)
                continue

            backoff = _BACKOFF_START
            batch = [self._build_message(raw, group) for raw in resp.messages]
            try:
                await handler(batch)
            except asyncio.CancelledError:
                return
            except Exception as exc:
                logger.error("queue-ti consumer: batch handler error: %s", exc)

    def _build_message(self, raw: queue_pb2.SubscribeResponse | queue_pb2.DequeueResponse, consumer_group: str = "") -> Message:
        group = consumer_group or self._consumer_group

        async def ack_fn() -> None:
            await self._stub.Ack(queue_pb2.AckRequest(id=raw.id, consumer_group=group))

        async def nack_fn(reason: str) -> None:
            await self._stub.Nack(queue_pb2.NackRequest(id=raw.id, error=reason, consumer_group=group))

        return Message(
            id=raw.id,
            topic=raw.topic,
            payload=bytes(raw.payload),
            metadata=dict(raw.metadata),
            created_at=_proto_ts_to_datetime(raw.created_at),
            retry_count=raw.retry_count,
            _ack_fn=ack_fn,
            _nack_fn=nack_fn,
        )
